mobile/application-manager-base.ts

```diff
@@ -27,29 +28,46 @@ export abstract class ApplicationManagerBase extends EventEmitter implements Mob
 		}).future<boolean>()();
 	}
 
+	private isChecking = false;
 	public checkForApplicationUpdates(): IFuture<void> {
 		return (() => {
-			let newInstalledApplications = this.getInstalledApplications().wait();
-			let previouslyInstalledApplications = this.lastInstalledApplications || [];
+			// As this method is called on 500ms, but it's execution may last much longer
+			// use locking, so the next executions will not get into the body, while the first one is still working.
+			// In case we do not break the next executions, we'll report each app as newly installed several times.
+			if(!this.isChecking) {
+				this.isChecking = true;
+				let currentlyInstalledApplications = this.getInstalledApplications().wait();
+				let previouslyInstalledApplications = this.lastInstalledApplications || [];
 
-			_(newInstalledApplications)
-				.difference(previouslyInstalledApplications)
-				.each(application => this.emit("applicationInstalled", application))
-				.value();
+				let newApplications = _.difference(currentlyInstalledApplications, previouslyInstalledApplications);
+				let removedApplications = _.difference(previouslyInstalledApplications, currentlyInstalledApplications);
 
-			_(previouslyInstalledApplications)
-				.difference(newInstalledApplications)
-				.each(application => this.emit("applicationUninstalled", application))
-				.value();
+				if(newApplications.length || removedApplications.length) {
+					this.applicationsLiveSyncStatus = _.filter(this.applicationsLiveSyncStatus, (app: Mobile.IApplicationLiveSyncStatus) => _.contains(removedApplications, app.applicationIdentifier));
+					this.getApplicationsLiveSyncSupportedStatus(newApplications).wait();
+				}
 
-			this.lastInstalledApplications = newInstalledApplications;
+				_.each(newApplications, application => this.emit("applicationInstalled", application, {isLiveSyncSupported: this.isLiveSyncSupported(application).wait()} ));
+				_.each(removedApplications, application => this.emit("applicationUninstalled", application));
+
+				this.lastInstalledApplications = currentlyInstalledApplications;
+				this.isChecking = false;
+			}
 		}).future<void>()();
 	}
 
+	public isLiveSyncSupported(appIdentifier: string): IFuture<boolean> {
+		return ((): boolean => {
```

Comment:
It appears you can safely skip returning `IFuture` here and go ahead and return `boolean` directly